mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -1650,13 +1665,40 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
         padGemmFilterTransform.passThrough("gemmN");
       }
 
-      TransformMapAttr padGemmFilterTransformAttr =
-          padGemmFilterTransform.get();
+      padGemmFilterTransformAttr = padGemmFilterTransform.get();
       gemmFilterPad = b.create<miopen::TransformOp>(loc, gemmFilter,
                                                     padGemmFilterTransformAttr);
       // filter pad end
     }
 
+    // KPack for filter tensor.
+    Value gemmFilterKPack = gemmFilterPad;
+
+    // FIXME. consider backward convolution.
+    if ((KPack > 1) && (convOpType == miopen::ConvOpType::Fwd)) {
+      BottomUpCTBuilder sourceTransform =
+          (isFilterPad) ? padGemmFilterTransform : filterTransform;
+      TransformMapAttr sourceTransformAttr =
+          (isFilterPad) ? padGemmFilterTransformAttr : filterTransformAttr;
+      Value source = (isFilterPad) ? gemmFilterPad : gemmFilter;
+
+      BottomUpCTBuilder kpackGemmFilterTransform =
+          BottomUpCTBuilder::above(sourceTransform, sourceTransformAttr);
+      // Passthrough gemmG (dim 0) and gemmM (dim 2).
+      kpackGemmFilterTransform.passThrough(sourceTransform.endName(0));
+      kpackGemmFilterTransform.passThrough(sourceTransform.endName(2));
+      // Use Unmerge to split gemmK (dim 1) into gemmK and gemmKPack, place
+      // gemmKPack at dim 3.
+      int64_t gemmKLength = sourceTransform.endSize(sourceTransform.endName(1));
```

Comment:
Nit: We could probably do `kName = sourceTransform.endName(1);` before this bit

Reply:
Will fix that in a separate cleanup PR.